src/lib/vasi-sync/src/atomic_tls_map.rs

```diff
@@ -0,0 +1,271 @@
+                        atomic::Ordering::Acquire,
+                        atomic::Ordering::Relaxed,
+                    )
+                    .is_ok()
+            })
+            .unwrap();
+        self.values[idx].get_mut().with(|table_value| {
+            let table_value = unsafe { &mut *table_value };
+            table_value.write(value)
+        });
+        unsafe { Ref::new(self.values[idx].get(), &self.refcounts[idx]) }
+    }
+
+    /// Retrieve the value associated with `key`, initializing it with `init` if `key`
+    /// is not already present.
+    ///
+    /// # Safety
+    ///
+    /// No other thread must access `self` with `key` in parallel, including
+    /// having outstanding references.
+    #[inline]
+    pub unsafe fn get_or_insert_with(&self, key: NonZeroUsize, init: impl FnOnce() -> V) -> Ref<V> {
+        let val = unsafe { self.get(key) };
+        val.unwrap_or_else(|| {
+            let val = init();
+            // SAFETY: Ensured by caller
+            unsafe { self.insert(key, val) }
+        })
+    }
+
+    /// Removes the value still for `key`, if any. Panics if this thread has
+    /// any outstanding references for `key`.
+    ///
+    /// # Safety
+    ///
+    /// *No thread* must have outstanding references to the value at `key`, or perform
+    /// parallel operations on `self` with `key`.
+    pub unsafe fn remove(&self, key: NonZeroUsize) -> Option<V> {
+        let idx = self.idx(key)?;
+        assert_eq!(self.refcounts[idx].get(), 0);
+        let value = self.values[idx].get_mut().with(|value| {
+            let value = unsafe { &mut *value };
+            unsafe { value.assume_init_read() }
+        });
+        // Syncs with `Acquire` on insertion
+        self.keys[idx].store(None, atomic::Ordering::Release);
+        Some(value)
+    }
+}
+
+impl<const N: usize, V, H> AtomicTlsMap<N, V, H>
+where
+    H: BuildHasher + Default,
+{
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self::new_with_hasher(Default::default())
+    }
+}
+
+impl<const N: usize, V, H> Drop for AtomicTlsMap<N, V, H>
+where
+    H: BuildHasher,
+{
+    fn drop(&mut self) {
+        for idx in 0..N {
+            // Relaxed because we have exclusive access to `self`. Any other threads
+            // that have accessed the table must have synchronized with the current
+            // thread to be able to drop `self` in the first place.
+            if self.keys[idx].load(atomic::Ordering::Relaxed).is_some() {
+                self.values[idx].get_mut().with(|value| {
+                    assert_eq!(self.refcounts[idx].get(), 0);
+                    // SAFETY: We have exclusive access to `self`.
+                    let value = unsafe { &mut *value };
+                    // SAFETY: We know the value is initialized.
+                    unsafe { value.assume_init_drop() }
+                })
+            }
+        }
+    }
+}
+
+pub struct Ref<'a, V> {
```

Comment:
static assert that `Ref<()>` is not `Send` or `Sync`, to make sure even if `V` is send or sync that `Ref` isn't?

Reply:
Done